src/lib/server/orders.ts

```diff
@@ -1594,44 +1668,76 @@ export async function updateAfterOrderPaid(order: Order, session: ClientSession)
 			productId: { $in: order.items.map((item) => item.product._id) }
 		})
 		.toArray();
-	for (const subscription of order.items.filter((item) => item.product.type === 'subscription')) {
-		const existingSubscription = subscriptions.find(
-			(sub) => sub.productId === subscription.product._id
-		);
+	const subscriptionProducts = order.items.filter((item) => item.product.type === 'subscription');
+	if (subscriptionProducts.length) {
+		const discounts = await collections.discounts
+			.find({
+				subscriptionIds: {
+					$in: order.items
+						.filter((item) => item.product.type === 'subscription')
+						.map((sub) => sub.product._id)
+				}
+			})
+			.toArray();
 
-		if (existingSubscription) {
-			const result = await collections.paidSubscriptions.updateOne(
-				{ _id: existingSubscription._id },
-				{
-					$set: {
-						paidUntil: add(max([existingSubscription.paidUntil, new Date()]), {
-							[`${runtimeConfig.subscriptionDuration}s`]: 1
-						}),
-						updatedAt: new Date(),
-						notifications: []
-					},
-					$unset: { cancelledAt: 1 }
-				},
-				{ session }
+		for (const subscription of subscriptionProducts) {
+			const existingSubscription = subscriptions.find(
+				(sub) => sub.productId === subscription.product._id
 			);
 
-			if (!result.modifiedCount) {
-				throw new Error('Failed to update subscription');
+			if (existingSubscription) {
+				const result = await collections.paidSubscriptions.updateOne(
+					{ _id: existingSubscription._id },
+					{
+						$set: {
+							paidUntil: add(max([existingSubscription.paidUntil, new Date()]), {
+								[`${runtimeConfig.subscriptionDuration}s`]: 1
+							}),
+							updatedAt: new Date(),
+							notifications: []
+						},
+						$unset: { cancelledAt: 1 }
+					},
+					{ session }
+				);
+
+				if (!result.modifiedCount) {
+					throw new Error('Failed to update subscription');
+				}
+			} else {
+				const freeProductsById: Record<string, { total: number; available: number; used: number }> =
+					{};
+
+				for (const discount of discounts) {
+					if (discount.mode === 'freeProducts' && discount.quantityPerProduct) {
+						for (const [productId, quantity] of Object.entries(discount.quantityPerProduct)) {
+							if (!freeProductsById[productId]) {
+								freeProductsById[productId] = {
+									total: 0,
+									available: 0,
+									used: 0
+								};
+							}
+							freeProductsById[productId].total += quantity;
+							freeProductsById[productId].available += quantity;
+						}
+					}
+				}
```

Comment:
This whole diff can be removed no? Since we don't care what happens after the order is paid, because the changes happen when the order is created

Reply:
ok but another problem probably, when you renew the subscription it should add new free products, currently the code doesn't do that (renewing the subcription doesn't add new products)